### ISSUE TITLE
Design matrix + constraints matrix doesn't work if no param_names are supplied.

### ISSUE THREAD

**Issue by picciama** [theislab/batchglm#102]
When supplying the design matrix as `numpy.ndarray` or `dask.array.core.Array`, there are no param_names created when parsing the supplied design matrix if they aren't supplied explicitly with the `param_names` argument.
https://github.com/theislab/batchglm/blob/5707d0a00da1be9485f48992bb14c6f9b295b712/batchglm/models/base_glm/utils.py#L34-L39

This is a problem when parsing the constraints matrix, where a list of `dmat_par_names` is necessary and cannot be empty.
Here:
https://github.com/theislab/batchglm/blob/5707d0a00da1be9485f48992bb14c6f9b295b712/batchglm/models/base_glm/utils.py#L73-L75
the `constraint_params` are assembled when parsing the supplied constraints matrix.
There is unfortunately no check whether `dmat_par_names is None`, which is the case for the above scenario (the `params` from above are `dmat_par_names` here).

We need to discuss how to resolve this:
- Do we really need `constraint_params` somewhere downstream or can we add a switch `if dmat_par_names is not None`?
- Or should we create default `param_names` instead of assigning `params = None` in the upper code excerpt?

**Comment by davidsebfischer** [comment]
I think we should force par_names to be given, either as column names in pandas or as `param_names` argument to https://github.com/theislab/batchglm/blob/5707d0a00da1be9485f48992bb14c6f9b295b712/batchglm/models/base_glm/utils.py#L17

alternatively we could also set default parameter names but i think that one would always want to have a mapping of coefficients to factors later, which is always done via the parameter names. if we force parameteres names to be defined in `parse_design` we do not have problems down stream, is that correct?

i would keep `constraint_params` because this is useful to understand the coefficient mappings.

**Comment by picciama** [comment]
> alternatively we could also set default parameter names but i think that one would always want to have a mapping of coefficients to factors later, which is always done via the parameter names. if we force parameteres names to be defined in `parse_design` we do not have problems down stream, is that correct?

Sure, `parse_constraints` is called directly after `parse_design` when creating an `InputDataGLM` object:
https://github.com/theislab/batchglm/blob/5707d0a00da1be9485f48992bb14c6f9b295b712/batchglm/models/base_glm/input.py#L94-L129

> i would keep `constraint_params` because this is useful to understand the coefficient mappings.

OK, I suggest in `input.InputDataGLM.__init__`:
https://github.com/theislab/batchglm/blob/5707d0a00da1be9485f48992bb14c6f9b295b712/batchglm/models/base_glm/input.py#L30-L32
We can leave the default `None` but add a switch 
```python
if isinstance(design_loc, np.ndarray) or isinstance(design_loc, dask.array.core.Array): 
    if not design_loc_names or len(design_loc_names) != design_loc.shape[1]:
        raise ValueError('Provide design_loc_names when passing design_loc ' \
            'as np.ndarray or dask.array.core.Array!')
```
I'll open a pull request OK?

**Comment by davidsebfischer** [comment]
Perfect, thanks!

**Comment by picciama** [comment]
I just noticed that there is no possibility to provide constraints as `pandas.DataFrame` or `patsy.design_info.DesignMatrix`.
I could add this functionality too?

**Comment by davidsebfischer** [comment]
> I just noticed that there is no possibility to provide constraints as `pandas.DataFrame` or `patsy.design_info.DesignMatrix`.
> I could add this functionality too?

constraints are always a dictionary before parsing or a ndarray / dask after parsing right? i dont think that `patsy.design_info.DesignMatrix` is appropriate in either stage, where did you want to use that? we could use `pandas.DataFrame` after parsing.

**Comment by picciama** [comment]
You are right. 
But you can also pass an existing constraints matrix when creating an `InputDataGLM` object. Let's leave it as it is. It is working after all.

**Comment by picciama** [comment]
No actually `InputDataGLM` always expects an existing contraints matrix or `None`.

**Comment by davidsebfischer** [comment]
Ok, let s leave it!

**Comment by picciama** [comment]
fixed by merge of #103
